asf_search/ASFStack.py

```diff
@@ -0,0 +1,276 @@
+
+        polarizations = list(set([product.properties['polarization'] for product in products]))
+        assert len(polarizations) == 1, 'All products must have the same polarization'
+
+        dates = [
+            datetime.strptime(product.properties['stopTime'], DATE_FMT) for product in products
+        ]
+        orbits = list(set([product.properties['orbit'] for product in products]))
+        pass_msg = 'All products must be from the same pass of a platform'
+        assert len(orbits) <= 2, pass_msg
+        if len(orbits) == 2:
+            assert max(orbits) - min(orbits) == 1, pass_msg
+            assert (max(dates) - min(dates)).minutes < 5, pass_msg
+
+        footprints = [geometry.shape(product.geometry) for product in products]
+        union = unary_union(footprints)
+        assert not union.is_empty, 'Products must overlap in space'
+
+    def __len__(self) -> int:
+        return len(self.products)
+
+    def __repr__(self):
+        return f'ASFProductGroup: orbit={self.orbit}, date={self.date}, n={self.__len__()}'
+
+
+class ASFStack:
+    """A sef of ASFProduct groups that form a valid InSAR stack. This means all groups:
+    - Are from the same relative orbit
+    - Have the same beam mode
+    - Contain at least one common polarization
+    - Overlap in space
+
+    Groups are sorted from earliest to latest date.
+
+    Parameters
+    ----------
+    `products`:
+        A list of ASFProducts from the same relative orbit.
+    """
+
+    def __init__(self, products: List[asf_search.ASFProduct]):
+        self.groups = self.group_products(products)
+        footprints = [group.footprint for group in self.groups]
+        self.union_footprint = unary_union(footprints).simplify(SIMPLIFY_TOL)
+        assert not self.union_footprint.is_empty, 'Groups must overlap in space'
+        self.intersect_footprint = intersection_all(footprints).simplify(SIMPLIFY_TOL)
+        self.start_date = min([group.date for group in self.groups])
+        self.end_date = max([group.date for group in self.groups])
+        self.dates = [group.date for group in self.groups]
+        self.relative_orbit = self.groups[0].relative_orbit
+
+    @staticmethod
+    def filter_by_orbit(products: List[asf_search.ASFProduct], orbit: int):
+        return [product for product in products if product.properties['orbit'] == orbit]
+
+    def group_products(self, products: List[asf_search.ASFProduct]):
+        """Group a set of products into a valid InSAR group."""
+        rel_orbits = list(set([product.properties['pathNumber'] for product in products]))
+        assert len(rel_orbits) == 1, 'All products must be from the same relative orbit'
+        abs_orbits = sorted(list(set([product.properties['orbit'] for product in products])))
+        product_groups = []
+        for i, orbit in enumerate(abs_orbits):
+            if i != 0 and orbit - abs_orbits[i - 1] == 1:
+                product_groups[-1] += self.filter_by_orbit(products, orbit)
```

Comment:
I think this removes the need for `i != 0` check

```suggestion
        product_groups = [self.filter_by_orbit(products, abs_orbits[0])]
        for i, orbit in enumerate(abs_orbits)[1:]:
            if orbit - abs_orbits[i - 1] == 1:
                product_groups[-1] += self.filter_by_orbit(products, orbit)
```

Reply:
Thanks!